/*
 * Copyright 2019 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto2";

package upsi;

import "upsi/crypto/elgamal.proto";

message TreeNodes {
	repeated TreeNode nodes = 1;
}

message TreeNode {
  repeated EncryptedElement elements = 2;
}

// Holds a set of encrypted values.
message EncryptedSet {
  repeated EncryptedElement elements = 1;
}

message HashSet {
  repeated bytes elements = 1;
}

// Holds an encrypted value and possibly encrypted associated data.
message EncryptedElement {
  optional ElGamalCiphertext element = 1; // ElGamal element (u, e)
  optional bytes payload = 3;
}

message PaillierCiphertext {
  optional bytes ciphertext = 1;
}

